# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def add_two(x):
    return x + 2

print(add_two)

<function add_two at 0x7fcbfc043820>


In [2]:
lambda x: x + 2

<function __main__.<lambda>(x)>

In [3]:
print(lambda x: x +2)

<function <lambda> at 0x7fcbfc043ca0>


In [4]:
# IIFE - immediatly invoked function expression
# Execute a lambda function immediatly after creation

(lambda x: x+2)(10)

12

#### Saving to a Variable

In [5]:
named_lambda = lambda y: y + 2

print(named_lambda)
named_lambda(15)

<function <lambda> at 0x7fcbfc043e50>


17

#### Multiple Inputs

In [7]:
def multiply(x,y):
    return x *y

multiply(5,10)

50

In [ ]:
(lambda x,y: x*y)(5,10)

In [9]:
## passing a function into a function


def apply_func(f,name):
    """
    f -> function to execute
    name -> name to execute the function on
    """
    return f(name)


def get_first_letter(string):
    return string[0]


apply_func(get_first_letter, 'Brian')

'B'

#### Passing a Lambda into a Function

In [12]:
# call the apply_func function, pass in lambda function as f argument, pass
apply_func(lambda string: string[-1], 'codingtemple')

'e'

In [13]:
#returning a function from a function

def create_multiplier(num):
    def inner_func(x):
        return num * x
    return inner_func(5)

create_multiplier(10)

50

#### Returning a Lambda from a Function

In [18]:
def new_multiplier(num):
    return lambda x: num * x

five = new_multiplier(5)
print(five)


three = new_multiplier(3)
print(three)

print(five(10), three(10))

<function new_multiplier.<locals>.<lambda> at 0x7fcbfcf6e700>
<function new_multiplier.<locals>.<lambda> at 0x7fcbfcf6edc0>
50 30


#### If Statements within Lambdas

In [19]:
def do_something(num):
    if num > 10:
        return num * 2
    else:
        return num + 2
    
do_something(25)

50

In [20]:
# Ternary: value_if_true if condition else value_if_false

def do_something_again(num):
    return num * 2 if num > 10 else num + 2

do_something(25)

50

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [45]:
def cube_arg(num):
    return lambda x: num ** x
cube = cube_arg(3)
    
    
print (cube(3))

27


In [27]:
cube_lambda = lambda x: x ** 3

print(cube_lambda)
cube_lambda(3)

<function <lambda> at 0x7fcbfcf8a790>


27

In [44]:
f = lambda x: x **3

f(3)

27

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [58]:
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [65]:
# function that we want to apply to each elelment in the list 
def squared_minus_two(number):
    return number**2 - 2


#random list
numbers = [12,4,5,1,7,3,9,11]


#Create an empty list
new_nums = []
#loop over the current number list
for num in numbers:
    #apply the function and save it to the variable
    new_num = squared_minus_two(num)
    #append the new numbers to our new list
    new_nums.append(new_num)
    
print(new_nums)

[142, 14, 23, -1, 47, 7, 79, 119]


In [66]:
# map(function, *iterable -> list, tuple, string, set, dict, etc.)

list(map(squared_minus_two, numbers))

[142, 14, 23, -1, 47, 7, 79, 119]

In [68]:
x_values = [1,2,3,4,5,6,7]
y_values = [10,12,25,23,54,21,53]


def multiply(x,y):
    print(f"x={x}, y={y}")
    return x * y


list(map(multiply, x_values, y_values))

x=1, y=10
x=2, y=12
x=3, y=25
x=4, y=23
x=5, y=54
x=6, y=21
x=7, y=53


[10, 24, 75, 92, 270, 126, 371]

In [69]:
x_values = [1,2,3,4,5,6,7]
y_values = [10,12,25]


def multiply(x,y):
    print(f"x={x}, y={y}")
    return x * y


list(map(multiply, x_values, y_values))

x=1, y=10
x=2, y=12
x=3, y=25


[10, 24, 75]

#### Using Lambda's with Map

In [70]:


numbers_again = [4,7,9,2,6,1,13]

list(map(lambda number: number**2 - 2, numbers_again))

[14, 47, 79, 2, 34, -1, 167]

In [72]:
x_vals = [1,2,3,4,5]
y_vals = [10,20,25,42,59]

list(map(lambda a,b: b/a, x_vals, y_vals))

[10.0, 10.0, 8.333333333333334, 10.5, 11.8]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number if it is even and halve it if it is odd</p>

In [78]:
numbers = range(1,11)


list(map(lambda x: x*2 if x % 2 == 0 else x/2, numbers))

[0.5, 4, 1.5, 8, 2.5, 12, 3.5, 16, 4.5, 20]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [79]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [85]:
#list of names

names = ['Bob', 'Andy', 'Allison', 'Amy', 'Briana']

def starts_with_a(name):
    if name[0].upper() =='A':
        return True
    else:
        return False


    
a_names_map = list(map(starts_with_a, names))
print(a_names_map)

[False, True, True, True, False]


In [86]:
a_names_filter = list(filter(starts_with_a, names))
print(a_names_filter)

['Andy', 'Allison', 'Amy']


In [87]:
def remainder(num):
    return num % 3

my_nums = [1,2,3,4,5,6,7,8,9,10]
list(map(remainder, my_nums))

[1, 2, 0, 1, 2, 0, 1, 2, 0, 1]

In [88]:
list(filter(remainder, my_nums))
#returns the values from the original list if the reuturn value from reminder(num) evaluates to True

[1, 2, 4, 5, 7, 8, 10]

In [89]:
for i in range(-10,10):
    print(i, bool(i))

-10 True
-9 True
-8 True
-7 True
-6 True
-5 True
-4 True
-3 True
-2 True
-1 True
0 False
1 True
2 True
3 True
4 True
5 True
6 True
7 True
8 True
9 True


#### Using Lambda's with Filter()

In [91]:
list(filter(lambda name: True if name[0].upper() == 'A' else False, names))

['Andy', 'Allison', 'Amy']

In [92]:
list(filter(lambda name: name[0].upper() == 'A', names))

['Andy', 'Allison', 'Amy']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [98]:
from random import randint
#create a random list of 20 random numbers between 1 and 100
random_nums = [randint(1,100) for _ in range(20)]

#print(random_nums)

numbers = [50, 57, 18, 85, 93, 56, 10, 35, 47, 81, 22, 29, 8, 93, 97, 10, 64, 15, 44, 88]


list(filter(lambda num: num >= 50, numbers))

[50, 57, 85, 93, 56, 81, 93, 97, 64, 88]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [100]:
from functools import reduce

help(reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, sequence[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence,
    from left to right, so as to reduce the sequence to a single value.
    For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the sequence in the calculation, and serves as a default when the
    sequence is empty.



In [101]:
reduce(lambda x,y: x+y, [1,2,3,4,5])

15

In [104]:
def add_nums(x,y):
    print('='*20)
    print(f"x: {x}")
    print(f"y: {y}")
    print(f"returns: {x + y}")
    return x + y

reduce(add_nums, [1,2,3,4,5])

x: 1
y: 2
returns: 3
x: 3
y: 3
returns: 6
x: 6
y: 4
returns: 10
x: 10
y: 5
returns: 15


15

In [ ]:
total = 0
for num in [1,2,3,4,5]:
    print(f"total: {total}, num: {num}")
    



#### Using Lambda's with Reduce()

In [107]:
list1 = [2,4,6,8,10]

In [105]:
result_lambda = reduce(lambda num1, num2: num1 * num2, [1,2,3,4,5])
print(result_lambda)

120


In [108]:
prod = 1
for num in list1:
    prod *=num
    
print(prod)

3840


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [ ]:
import time

def capitalize(name):
    # takes in a name, upper cases the first letter, lower cases the rest
    return name[0].upper() + name[1:].lower()

def full_name(first, last):
    time.sleep(5)
    #takes in a firt name, calls the capitalize on first, takes in last name, calls capitalize on last
    #and return concatenated string of first + space + last
    return capitalize(first) + ' ' + capitalize(last)
    
    
def greet_person(first_name, last_name, day_of_week):
    #takes in first_name, last_name, and a day of the week
    print('Hello')
    #sets the person_full_name variable to the return value when full_name function is called by passing in first_name, last_name
    person_full_name = full_name(first_name, last_name)
    print(f"Hello {person_full_name}, it sure is beautiful on this {capitalize(day_of_week)}")
    
    
greet_person('han', 'solo', 'tuesday')

In [111]:
#base case

def add_nums(num):
    if num <= 1:
        print(f"add_nums({num}) = {num}")
        return num
    
    else:
        print(f"add_nums({num}) = {num} + add_nums({num -1})")
        return num + add_nums(num - 1)

    
add_nums(5)
        

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1


15

In [ ]:
# a recursive function is a function that calls for itself

#### Writing a Factorial Function

In [112]:
#5! = 5*4!
#4! = 4*3!
#3! = 3*2!
#2! = 2*1!
#1! = 1


def factorial(num):
    output = 1
    while num > 0:
        output *= num
        num -= 1
    return output


factorial(5)

120

In [114]:
def factorial(num):
    # set up base case that when num is 1 or less it returns num
    if num <= 1:
        return num
    else:
        return num * factorial(num -1)
    #otherwise the factorial of num = num * factorial(num -1)
    
factorial(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [122]:
def sub_nums(num):
    if num <= 1:
        print(f"sub_nums({num}) = {num}")
        return num
    
    else:
        print(f"sub_nums({num}) = {num} - sub_nums({num -1})")
        return num - sub_nums(num - 1)

    
sub_nums(5)

sub_nums(5) = 5 - sub_nums(4)
sub_nums(4) = 4 - sub_nums(3)
sub_nums(3) = 3 - sub_nums(2)
sub_nums(2) = 2 - sub_nums(1)
sub_nums(1) = 1


3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

In [123]:
lengths = map(len, ['Brian', 'Stanton', 'Coding Temple'])

print(lengths)

In [124]:
for l in lengths:
    print(l)

5
7
13


In [125]:
for l in lengths:
    print(l)

In [126]:
a_new_list = [2,3,4,5,6,7,8]

list_iter = iter(a_new_list)
print(list_iter)

In [128]:
next(list_iter)

3

In [ ]:
for x in list_iter:
    print(x)

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [ ]:
def my_range(start, stop, step=1):
    while start < stop:
        return start
        start += step
        
        
my_range(0,10)    

In [38]:
def my_range(start, stop, step=1):
    while start < stop:
        yield start
        start += step
        
my_range(0,10)

<generator object my_range at 0x7fb8d474d660>

In [40]:
for x in my_range(0,10):
    print(x)

0
1
2
3
4
5
6
7
8
9


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

In [ ]:
def infinite(x):
    while True:
    yield x
    x += 1
    
    
infinite_gen = infinite(1)

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [57]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

print(type(places))

def filter_empty(place):
    empty = ["Argentina", "San Diego", "Boston", "New York"]
    return True if place in empty else False

filtered_empty = filter(filter_empty, places)

empty = list(filtered_empty)
print(empty)

<class 'list'>
['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [27]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

#define function # define parameter
    #sort list by last name
        #do something here that returns a result
        
author.sort(key = lambda name: name.split(" ")[-1].lower())
print(author)
       
        
#call the function and insert variable/list
#help(list.sort)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [72]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

print(type(places))

<class 'list'>


In [75]:
list(map(lambda x: (x*9/5) + 32, places))

TypeError: unsupported operand type(s) for /: 'tuple' and 'int'

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [123]:
#fibonacci sequence - F(n) = F(n-1) + F(n-2)


#fib(5) = 8
def fib(x):
    if x in {0,1}:
        return x
    else:
        return (fib(x-1) + fib(x-2))
    
num = 5
    
for i in range(num):
       print(fib(i))

0
1
1
2
3


In [112]:
#for i in range(n):
#    print(f"Iteration {i}: {fib(i)}")

#### Exercise #5 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [63]:
def num_gen(n):
    for x in range(n):
        yield x**2
        
        
for x in num_gen(5):
    print(x)

0
1
4
9
16
